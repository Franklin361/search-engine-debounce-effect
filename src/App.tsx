import { Form } from "./components/Form"
import { useEffect, useState } from 'react';
import { searchPokemon } from "./utils/searchPokemon";
import { ResponseAPI } from "./interface/pokemon";
import { Pokemon } from "./components/Pokemon";

const delay = 1000; // 1s

const App = () => {

  const [pokemon, setPokemon] = useState<ResponseAPI | null>({} as ResponseAPI);
  const [isLoading, setIsLoading] = useState(false)

  const [value, setValue] = useState('');
  const onChange = (e: React.ChangeEvent<HTMLInputElement>) => setValue(e.target.value);

  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(() => {

    const timer = setTimeout(() => setDebouncedValue(value), delay || 500)

    return () => clearTimeout(timer)
  }, [value, delay]);

  useEffect(() => {
    const controller = new AbortController();
    if (debouncedValue) {
      setIsLoading(true)
      searchPokemon(debouncedValue, controller.signal)
        .then(data => {
          setPokemon(data);
          setIsLoading(false);
        })
    }

    return () => controller.abort();
  }, [debouncedValue])


  return (
    <div className="container">
      <h1> <span>Search Engine</span> whit <span>Debounce Effect</span> </h1>
      <Form {...{ value, onChange }} />
      {
        isLoading 
          ? <span>Loading Results...</span>
          : <Pokemon pokemon={pokemon}/>
      }
      
    </div>
  )
}
export default App