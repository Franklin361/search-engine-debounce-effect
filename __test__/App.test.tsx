import { render } from '@testing-library/react';
import App from '../src/App';

describe('ðŸ“‚ App.tsx', () => {
    test('ðŸŸ£ Should show title', async () => {
        const { container } = render(<App />)
        expect(container.textContent).toContain('Search Engine')
    });
    test('ðŸŸ£ Should match the snapshot', async () => {
        const { container } = render(<App />)
        expect(container).toMatchSnapshot();
    });

})