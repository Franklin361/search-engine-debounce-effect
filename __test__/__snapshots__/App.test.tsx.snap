// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ðŸ“‚ App.tsx ðŸŸ£ Should match the snapshot 1`] = `
<div>
  <main>
    <h1>
       
      <span>
        Search Engine
      </span>
       whit 
      <span>
        Debounce Effect
      </span>
       
    </h1>
    <label
      for="pokemon"
    >
      Name or ID of a Pokemon
    </label>
    <input
      id="pokemon"
      placeholder="Example: Pikachu"
      type="text"
      value=""
    />
  </main>
</div>
`;
